index.d.ts

```diff
@@ -1,74 +1,92 @@
 
-  export type LinkTargetResolver = (uri: string, text: string, title?: string) => string
+  type ReferenceType = mdast.ReferenceType
 
-  type Renderer<T> = (props: T) => ReactElement<T>
+  type LinkTargetResolver = (uri: string, text: string, title?: string) => string
+
+  type Renderer<T> = (props: T) => ElementType<T>
   interface Renderers {
     [key: string]: string | Renderer<any>
   }
 
-  export interface ReactMarkdownProps {
+  interface ReactMarkdownPropsBase {
     readonly className?: string
-    readonly source?: string
-    readonly children?: string
     readonly sourcePos?: boolean
     readonly includeNodeIndex?: boolean
     readonly rawSourcePos?: boolean
-    readonly escapeHtml?: boolean
-    readonly allowDangerousHtml?: boolean
-    readonly skipHtml?: boolean
-    readonly allowNode?: (node: Content, index: number, parent: NodeType) => boolean
-    readonly allowedTypes?: NodeType[]
-    readonly disallowedTypes?: NodeType[]
+    readonly allowNode?: (node: mdast.Content, index: number, parent: NodeType) => boolean
     readonly linkTarget?: string | LinkTargetResolver
     readonly transformLinkUri?:
       | ((uri: string, children?: ReactNode, title?: string) => string)
       | null
     readonly transformImageUri?:
       | ((uri: string, children?: ReactNode, title?: string, alt?: string) => string)
       | null
-    readonly unwrapDisallowed?: boolean
-    readonly renderers?: {[nodeType: string]: ReactType}
+    readonly renderers?: {[nodeType: string]: ElementType}
+    /** @deprecated use plugins */
     readonly astPlugins?: PluggableList
     readonly plugins?: PluggableList
   }
 
-  export var types: NodeType[]
-  export var renderers: Renderers
-  export var uriTransformer: (uri: string) => string
+  interface SourceProp {
+    /** @deprecated use children */
+    readonly source: string
+  }
+
+  interface ChildrenProp {
+    readonly children: string
+  }
+
+  interface AllowedTypesProp {
+    readonly allowedTypes?: NodeType[]
+  }
+
+  interface DisallowedTypesProp {
+    readonly disallowedTypes: NodeType[]
+    readonly unwrapDisallowed?: boolean
+  }
+
+  interface EscapeHtmlProp {
+    /** @deprecated use allowDangerousHtml */
+    readonly escapeHtml?: boolean
+    readonly allowDangerousHtml?: boolean
+  }
+
+  interface SkipHtmlProp {
+    readonly skipHtml?: boolean
+  }
+
+  type ReactMarkdownProps = ReactMarkdownPropsBase &
+    ((SourceProp & Not<ChildrenProp>) | (ChildrenProp & Not<SourceProp>)) &
+    (
+      | (AllowedTypesProp & Not<DisallowedTypesProp>)
+      | (DisallowedTypesProp & Not<AllowedTypesProp>)
+    ) &
+    ((EscapeHtmlProp & Not<SkipHtmlProp>) | (SkipHtmlProp & Not<EscapeHtmlProp>))
+
+  const types: NodeType[]
+  const renderers: Renderers
+  function uriTransformer(uri: string): string
 }
 
-declare class ReactMarkdown extends Component<ReactMarkdown.ReactMarkdownProps, {}> {}
+declare function ReactMarkdown(props: ReactMarkdown.ReactMarkdownProps): ReactElement
```

Comment:
this is roughly the equivalent of
```ts
declare const ReactMarkdown: React.FunctionComponent<ReactMarkdownProps>
```

However, `var`/`let`/`const` cannot have an overlapping name with `declare namespace`, while `function` does support it.

Reply:
the previous `declare class` is mismatches with the actual implementation
https://github.com/remarkjs/react-markdown/blob/4f5dbe21d0375f08e44f453e41505f2b45c1230e/src/react-markdown.js#L19